### DISCUSSION TITLE
Hero meeting screen, "fast move army" buttons should transfer creatures in respect to their power

### DISCUSSION THREAD

**Discussion by Branikolog** [ihhub/fheroes2#4396] [Ideas]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

Currently, move army buttons move the entire army to a meeting hero starting from the left slot up to the right one (also this feature unites stack of the same type, if there are any):
![image](https://user-images.githubusercontent.com/55348946/136847335-3f5cec9e-fa63-4e6e-b065-2804ae237251.png)
![image](https://user-images.githubusercontent.com/55348946/136847356-d10060a7-6301-4407-89ee-15215a28bcff.png)

### Describe the possible solution

This logic is inconvenient. The most common usage of this button is to give army from one hero to another making the strongest possible army. 
For example: we should transfer Titans first, then Rocs, then Boars and the last one (if we have free space for them) Halflings.
So in previous screenshots the result should be: 
![image](https://user-images.githubusercontent.com/55348946/136846750-ae94aa93-f550-493f-92ff-54268f21d953.png)
![image](https://user-images.githubusercontent.com/55348946/136847293-11a6f915-e27b-43b2-92b9-03dcc793e59a.png)

**Comment by ihhub** [comment]
I absolutely disagree with this. What if my first hero needs to have maximum move points?

**Comment by Branikolog** [comment]
> I absolutely disagree with this. What if my first hero needs to have maximum move points?

With the current logic this button leaves not the fastest troop in hero army, but the one, that is placed in the most right slot. Anyway, do you prefer leaving 1 Phoenix for your secondary hero, while transferring army of Sorceress castle?

**Comment by Branikolog** [comment]
Hmmm...
The only advantage current logic gives - we can manually change the outcome of transferring of all units with this button, by changing the army order BEFORE pressing the button. I think it's quite uncomfortable.... As rearranging of army before moving it... We can move the army manually then... 
But, I've just caught an idea:
We should implement the proposed logic of leaving "1" of the weakest (as the most common to use), but we can change it by selecting a troop, we want to leave the last in hero's army. (click on it and then push the button for fast transferring: all army will be transferred, leaving the whole stack of the selected troop (in a case there's no place for such type of troop, or leaving "1" of the selected stack in a case there's a room for every troop in target hero)

**Comment by LeHerosInconnu** [comment]
Hello @Branikolog and @ihhub,

I think the default action should be to keep one creature that is the fastest (in priority) AND the weakest.
Generally, this is what I do and what the majority of players I've observed in let's play do, which is normal because it's better to have a fast (at least temporarily) primary/secondary hero.

For the selection of the creature/troop to keep, I made an enhancement proposal in this way here: https://github.com/ihhub/fheroes2/discussions/3742. :)

Both enhancements should be combined and active at the same time.

**Comment by Branikolog** [comment]
Hello, @LeHerosInconnu 

> the fastest (in priority) AND the weakest.

We cannot combine both of this conditions. Also we cannot make this logic obvious and predictable for a player: the type of creatures being transferred will be different in each case. For example: if we transfer black dragons and cyclops, what creature should stay in initial hero? The fastest but ultimate dragon? Or weak and useless cyclops? From my point - we should leave weakest creature in initial hero: this is always predictable and suits the overall purpose of the button - fast transferring from a weaker hero to a stronger one.
The screenshots above are not the most demonstrative, which caused our dissension. :) I agree, that in the situation on a screenshots everyone would leave boars for sure in the initial hero army. But this case is considered as exception.
Here's another, and the most common case:
![image](https://user-images.githubusercontent.com/55348946/136926634-49fb1a18-f041-42b7-a281-93c3eda2de35.png)
Do we need to leave fast Liches, rather than skeletons in barbarian Atlas' army? Definitely not.


> For the selection of the creature/troop to keep, I made an enhancement proposal in this way here: #3742. :)

I definitely agree with your proposal here. I'm sorry I forgot about your thread. My idea 100% coincides with your proposal. :)

**Comment by ihhub** [comment]
I feel we have to make it as a discussion as I don't see that we cover 100% of cases with your proposed logic and for players it's not obvious.

**Comment by Branikolog** [comment]
@ihhub, sure!
We need to hear everyone's opinion on the issue, as it discovered to be not such simple.
You can move this thread.

To summarize my point:
- Current implementation is uncomfortable, due to compelled rearranging of the initial army in order to be sure, that needed trop will take the most right slot. While transferring armies they shouldn't change their arrangement and at the same time we shouldn't make player intendedtively place certain troop in the most right slot.

- The proposed solution of transferring from the strongest creature to the weakest is 99% predictable and the most common in use.
With the combination of nice proposal made by @LeHerosInconnu (#3742) this logic will handle the most cases player can expect from this feature.

**Comment by LeHerosInconnu** [comment]
@Branikolog and @ihhub.

> > the fastest (in priority) AND the weakest.

This means that if two troops are composed of a creature of the same speed the weaker one should be kept.

> Here's another, and the most common case:
> ![image](https://user-images.githubusercontent.com/55348946/136926634-49fb1a18-f041-42b7-a281-93c3eda2de35.png) Do we need to leave fast Liches, rather than skeletons in barbarian Atlas' army? Definitely not.

It depends on the goal, if afterwards Atlas should be as mobile as possible or not, or if Rebbecca should be as powerful as possible or not. 

> > For the selection of the creature/troop to keep, I made an enhancement proposal in this way here: #3742. :)
> 
> I definitely agree with your proposal here. I'm sorry I forgot about your thread. My idea 100% coincides with your proposal. :)

Yes. No worries. :)

> I feel we have to make it as a discussion as I don't see that we cover 100% of cases with your proposed logic and for players it's not obvious.

I think it is not possible to cover 100% of the cases.

As I said, generally the player wants the secondary hero to have as much mobility as possible.
The main hero is there principally to combat, so his army should be strong even if he is a little less mobile.
But the main hero, may have to stay very mobile when he is not in a combat situation.
It is therefore complicated to determine the player's goal when he transfers his armies between heroes.

So you would have to make a choice in the default action, either to focus on the hero's movement speed, or to focus on the power of the hero's army.

This can eventually be set in the fheroes2.cfg file:
"# meeting troops transfer: 1 - 2 (1 for hero speed - 2 for hero power)"

**Comment by Branikolog** [comment]
@LeHerosInconnu 

> I think it is not possible to cover 100% of the cases.

100% agreed. :)
Me personally consider the situation we're talking about as "Army transferring", so the main goal is to transfer as much army, as we can. (Not making an initial hero as mobile, as possible (which is often a secondary purpose.) ) Which means, that the strongest creatures should be transferred and we weakest should be left in initial army.

> It is therefore complicated to determine the player's goal when he transfers his armies between heroes.

That's true. We cannot read players thoughts, but we should aim to allow player translate this thoughts into reality as much predictable and obvious, as we can.

I think it's impossible to implement such logic, which chooses the most mobile and the less valuable creature to be left within this issue, because it is super-complex and still cannot reveal players goals for sure.

**Comment by LeHerosInconnu** [comment]
@Branikolog.

> Me personally consider the situation we're talking about as "Army transferring", so the main goal is to transfer as much army, as we can. (Not making an initial hero as mobile, as possible (which is often a secondary purpose.) ) Which means, that the strongest creatures should be transferred and we weakest should be left in initial army.

Here is what I have noticed (and what I also practice in most cases).
Generally when a player (in the let's plays I've watched) performs this action of transferring his troops from a secondary hero to the main hero (even without using the arrows or other system at his disposal, so moving the troops one by one), he tries to have the most powerful army for his main hero AND AT THE SAME TIME he tries to have the secondary hero with the highest movement speed.
And when a player transfers his troops from the main hero to a secondary hero (no combat in perspective for example), he tries to have the main hero with the highest movement speed.
What we notice here is that the player makes sure that his heroes keep the highest possible movement speed in all cases.

**Reply by Branikolog** [reply]
@LeHerosInconnu 

I have to agree, that during gameplay we often leave fast troops in initial army.

But, Heroes2 has not so big creature pool. I'm pretty sure you will never leave 1 Phoenix in initial hero of sorceress faction. As well as 1 Power Lich while meeting within Necromancers heroes, which are the the fastest creatures of these factions. Also I doubt, that you would like to reduce your Champion (not so strong, though) stack in order to make secondary hero faster (In some cases probably you do, but not so often, as Champions are still quite valuable, because other troops are weak too. :-P ).
But anyway, to handle such non-common (in my opinion) cases we would add a selection mark to set lowest transferring priority. And for the rest cases - we prioritize strongest creatures for transferring. What do you think?

I see few disadvantages of your proposal:

- While transferring for example Wizard faction everything seems clear: Titans, Arch Magi and boars have same speed and we leave weakest boars on transferring in the initial army, which is ok. But after a battle we can loose boars or even Arch Magi. So, in this case the button would leave 1 Titan on transferring?

- Leaving the fastest troop as Titan (Or any other creature with the highest speed and high value) is actual for main heroes. You master much more secondary heroes, comparing to main ones. So this logic will fit not so often.

- I'm still convinced, that value prioritizing is more predictable and obvious: we always leave, for example, hobbits in initial army. If they die - then boars, etc. This is more predictable, than considering speed of the creatures. The first thing you learn about each faction - is the order of creatures, regarding their value or strength. Not speed. So this base knowledge makes suggested proposal more predictable, rather, than speed prioritizing.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> @LeHerosInconnu
> 
> I have to agree, that during gameplay we often leave fast troops in initial army.
> 
> But, Heroes2 has not so big creature pool. I'm pretty sure you will never leave 1 Phoenix in initial hero of sorceress faction. As well as 1 Power Lich while meeting within Necromancers heroes, which are the the fastest creatures of these factions. Also I doubt, that you would like to reduce your Champion (not so strong, though) stack in order to make secondary hero faster (In some cases probably you do, but not so often, as Champions are still quite valuable, because other troops are weak too. :-P ). But anyway, to handle such non-common (in my opinion) cases we would add a selection mark to set lowest transferring priority. And for the rest cases - we prioritize strongest creatures for transferring. What do you think?

Actually, I use Champion, Wolf, Grand Elf, Power Lich or Vampire Lord, Gargoyle, Boar, Nomad, Genie, eventually Phoenix (I've had secondary hero(es) with one Black Dragon, having 9 or 10 Black Dragons in the main hero's army doesn't make too much difference :)), or the fastest troop I have for my secondary heroes' armies (scoot, resource gatherer, fresh troop conveyor, etc.), and if the heroes are exposed to the opponents, sometimes with two, three or five one-creature troops (so the hero can take a few casualties before he can run away or surrender).
You can even see this easily by loading one of the many save files I provide for the issues. :)

The mobility of the heroes should to be neglected in Heroes 2, it is an important point especially in the high difficulty levels.
And having one more Boar and one less Halfling in a combat army is not what will determine the outcome of the combat in most cases.
Whereas a secondary hero who can move three more squares each turn on the adventure map with a Boar instead of a Halfling in his army can bring much more advantages.

> I see few disadvantages of your proposal:
> 
> * While transferring for example Wizard faction everything seems clear: Titans, Arch Magi and boars have same speed and we leave weakest boars on transferring in the initial army, which is ok. But after a battle we can loose boars or even Arch Magi. So, in this case the button would leave 1 Titan on transferring?

It really depends on the situation, but this kind of case seems to me more marginal and the player can easily remedy it by adjusting the result obtained when using the troop transfer action.
Troop transfer with the need for fast heroes is much more frequent in my opinion.
Look at some let's plays with experienced players and you will see the need for fast heroes.

> * Leaving the fastest troop as Titan (Or any other creature with the highest speed and high value) is actual for main heroes. You master much more secondary heroes, comparing to main ones. So this logic will fit not so often.
> * I'm still convinced, that value prioritizing is more predictable and obvious: we always leave, for example, hobbits in initial army. If they die - then boars, etc. This is more predictable, than considering speed of the creatures. The first thing you learn about each faction - is the order of creatures, regarding their value or strength. Not speed. So this base knowledge makes suggested proposal more predictable, rather, than speed prioritizing.

The speed of the creatures is also very important during a combat because it determines who will play first.
Also, a secondary hero (or even the main hero attacked by surprise without his strong army), will have a better chance to retreat or surrender if he has one Boar in his army rather than one Halfling.
